File: rust/src/problems/p763.rs
pub struct Solution;

impl Solution {
    pub fn partition_labels(s: String) -> Vec<i32> {
        let mut result = vec![];
        let mut last = [0; 26];
        let s = s.as_bytes();
        for (i, c) in s.iter().enumerate() {
            let idx = (c - b'a') as usize;
            last[idx] = i;
        }
        let mut start = 0;
        let mut end = 0;
        for (i, c) in s.iter().enumerate() {
            let idx = (c - b'a') as usize;
            end = end.max(last[idx]);
            if i == end {
                result.push((end - start + 1) as i32);
                start = end + 1;
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        assert_eq!(
            vec![9, 7, 8],
            Solution::partition_labels("ababcbacadefegdehijhklij".to_string())
        );
    }

    #[test]
    fn case2() {
        assert_eq!(
            vec![10],
            Solution::partition_labels("eccbbbbdec".to_string())
        );
    }
}

File: cpp/p763.cpp
#include <array>
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> partitionLabels(string s) {
        array<int, 26> last;
        last.fill(-1);
        for (int i = 0; i < s.size(); i++) {
            last[s[i] - 'a'] = i;
        }
        vector<int> ans;
        for (int start = 0, end = 0; start < s.size();) {
            for (int j = start; j <= end; j++) {
                int k = last[s[j] - 'a'];
                end = max(end, k);
            }
            ans.push_back(end - start + 1);
            end += 1;
            start = end;
        }
        return ans;
    }
};

int main() {
    auto r = Solution().partitionLabels("ababcbacadefegdehijhklij");
    vector<int> expected = {9, 7, 8};
    assert(r == expected);
    return 0;
}